File: backend/internal/services/student_service.go
package services

import (
	"backend/internal/models"
	"backend/internal/repository"
	"fmt"
	"strings"
)

type StudentService interface {
	// Basic CRUD operations
	CreateStudent(req models.CreateStudentRequest) (*models.StudentResponse, error)
	GetStudentByID(id uint) (*models.StudentResponse, error)
	GetStudentByUserID(userID uint) (*models.StudentResponse, error)
	GetStudents(filters repository.StudentFilters) ([]models.StudentResponse, int64, error)
	UpdateStudent(studentID uint, updates map[string]interface{}) (*models.StudentResponse, error)
	DeleteStudent(studentID uint) error

	// Business specific operations
	GetStudentsByBusiness(businessID uint, filters repository.StudentFilters) ([]models.StudentResponse, int64, error)
	GetActiveStudentsByBusiness(businessID uint) ([]models.StudentResponse, error)
	GetInactiveStudentsByBusiness(businessID uint) ([]models.StudentResponse, error)

	// Status operations
	ChangeStudentStatus(studentID uint, status int) error
	GetActiveStudents() ([]models.StudentResponse, error)
	GetInactiveStudents() ([]models.StudentResponse, error)

	// Search
	SearchStudents(searchTerm string, limit int, businessID ...uint) ([]models.StudentResponse, error)
	SearchStudentsByBusiness(businessID uint, searchTerm string, limit int) ([]models.StudentResponse, error)

	// Statistics
	GetStudentStats(businessID ...uint) (map[string]interface{}, error)
	GetGuardianStats(businessID ...uint) (map[string]interface{}, error)

	// Bulk operations
	BulkUpdateStudentStatus(studentIDs []uint, status int) error

	// Validation
	ValidateCreateStudentRequest(req models.CreateStudentRequest) error
	ValidateUpdateStudentRequest(req models.UpdateStudentRequest) error
}

type studentService struct {
	studentRepo  repository.StudentRepository
	userRepo     repository.UserRepository
	businessRepo repository.BusinessRepository
}

func NewStudentService(studentRepo repository.StudentRepository, userRepo repository.UserRepository, businessRepo repository.BusinessRepository) StudentService {
	return &studentService{
		studentRepo:  studentRepo,
		userRepo:     userRepo,
		businessRepo: businessRepo,
	}
}

func (s *studentService) CreateStudent(req models.CreateStudentRequest) (*models.StudentResponse, error) {
	// Validate request
	if err := s.ValidateCreateStudentRequest(req); err != nil {
		return nil, err
	}

	// Check if user exists and is not already a student
	if _, err := s.userRepo.GetByID(req.UserID); err != nil {
		return nil, fmt.Errorf("user not found")
	}

	// Check if user is already a student
	exists, err := s.studentRepo.StudentUserExists(req.UserID)
	if err != nil {
		return nil, fmt.Errorf("failed to check if user is already a student")
	}
	if exists {
		return nil, fmt.Errorf("user is already a student")
	}

	// Check if business exists
	_, err = s.businessRepo.GetByID(req.BusinessID)
	if err != nil {
		return nil, fmt.Errorf("business not found")
	}

	// Initialize information if nil
	if req.Information == nil {
		req.Information = make(models.JSONB)
	}

	// Create student
	student := &models.Student{
		Name:           req.Name,
		UserID:         req.UserID,
		BusinessID:     req.BusinessID,
		GuardianName:   req.GuardianName,
		GuardianNumber: req.GuardianNumber,
		GuardianEmail:  req.GuardianEmail,
		Information:    req.Information,
		Status:         1, // Active by default
	}

	if err := s.studentRepo.Create(student); err != nil {
		return nil, fmt.Errorf("failed to create student: %v", err)
	}

	// Get student with relations
	studentWithRelations, err := s.studentRepo.GetStudentWithRelations(student.ID)
	if err != nil {
		return nil, fmt.Errorf("failed to get created student")
	}

	return s.toStudentResponse(studentWithRelations), nil
}

func (s *studentService) GetStudentByID(id uint) (*models.StudentResponse, error) {
	student, err := s.studentRepo.GetStudentWithRelations(id)
	if err != nil {
		return nil, fmt.Errorf("student not found")
	}

	return s.toStudentResponse(student), nil
}

func (s *studentService) GetStudentByUserID(userID uint) (*models.StudentResponse, error) {
	student, err := s.studentRepo.GetByUserID(userID)
	if err != nil {
		return nil, fmt.Errorf("student profile not found")
	}

	// Get with relations
	studentWithRelations, err := s.studentRepo.GetStudentWithRelations(student.ID)
	if err != nil {
		return nil, fmt.Errorf("failed to get student details")
	}

	return s.toStudentResponse(studentWithRelations), nil
}

func (s *studentService) GetStudents(filters repository.StudentFilters) ([]models.StudentResponse, int64, error) {
	// Set default pagination
	if filters.Page == 0 {
		filters.Page = 1
	}
	if filters.Limit == 0 {
		filters.Limit = 10
	}

	students, total, err := s.studentRepo.GetAllWithRelations(filters)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to get students: %v", err)
	}

	var responses []models.StudentResponse
	for _, student := range students {
		responses = append(responses, *s.toStudentResponse(&student))
	}

	return responses, total, nil
}

func (s *studentService) UpdateStudent(studentID uint, updates map[string]interface{}) (*models.StudentResponse, error) {
	// Get existing student
	student, err := s.studentRepo.GetByID(studentID)
	if err != nil {
		return nil, fmt.Errorf("student not found")
	}

	// Update fields
	if name, ok := updates["name"]; ok {
		if nameStr, ok := name.(string); ok && nameStr != "" {
			student.Name = nameStr
		}
	}

	if guardianName, ok := updates["guardian_name"]; ok {
		if guardianNameStr, ok := guardianName.(string); ok {
			student.GuardianName = guardianNameStr
		}
	}

	if guardianNumber, ok := updates["guardian_number"]; ok {
		if guardianNumberStr, ok := guardianNumber.(string); ok {
			student.GuardianNumber = guardianNumberStr
		}
	}

	if guardianEmail, ok := updates["guardian_email"]; ok {
		if guardianEmailStr, ok := guardianEmail.(string); ok {
			student.GuardianEmail = guardianEmailStr
		}
	}

	if information, ok := updates["information"]; ok {
		if infoMap, ok := information.(models.JSONB); ok {
			student.Information = infoMap
		}
	}

	if status, ok := updates["status"]; ok {
		if statusInt, ok := status.(int); ok && (statusInt == 0 || statusInt == 1) {
			student.Status = statusInt
		}
	}

	// Save updates
	if err := s.studentRepo.Update(student); err != nil {
		return nil, fmt.Errorf("failed to update student: %v", err)
	}

	// Get updated student with relations
	updatedStudent, err := s.studentRepo.GetStudentWithRelations(student.ID)
	if err != nil {
		return nil, fmt.Errorf("failed to get updated student")
	}

	return s.toStudentResponse(updatedStudent), nil
}

func (s *studentService) DeleteStudent(studentID uint) error {
	// Check if student exists
	_, err := s.studentRepo.GetByID(studentID)
	if err != nil {
		return fmt.Errorf("student not found")
	}

	if err := s.studentRepo.Delete(studentID); err != nil {
		return fmt.Errorf("failed to delete student: %v", err)
	}

	return nil
}

func (s *studentService) GetStudentsByBusiness(businessID uint, filters repository.StudentFilters) ([]models.StudentResponse, int64, error) {
	// Set default pagination
	if filters.Page == 0 {
		filters.Page = 1
	}
	if filters.Limit == 0 {
		filters.Limit = 10
	}

	students, total, err := s.studentRepo.GetByBusinessID(businessID, filters)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to get students by business: %v", err)
	}

	var responses []models.StudentResponse
	for _, student := range students {
		// Get with relations for complete data
		studentWithRelations, err := s.studentRepo.GetStudentWithRelations(student.ID)
		if err != nil {
			continue // Skip this one if there's an issue
		}
		responses = append(responses, *s.toStudentResponse(studentWithRelations))
	}

	return responses, total, nil
}

func (s *studentService) GetActiveStudentsByBusiness(businessID uint) ([]models.StudentResponse, error) {
	students, err := s.studentRepo.GetActiveStudentsByBusiness(businessID)
	if err != nil {
		return nil, fmt.Errorf("failed to get active students: %v", err)
	}

	var responses []models.StudentResponse
	for _, student := range students {
		studentWithRelations, err := s.studentRepo.GetStudentWithRelations(student.ID)
		if err != nil {
			continue
		}
		responses = append(responses, *s.toStudentResponse(studentWithRelations))
	}

	return responses, nil
}

func (s *studentService) GetInactiveStudentsByBusiness(businessID uint) ([]models.StudentResponse, error) {
	students, err := s.studentRepo.GetInactiveStudentsByBusiness(businessID)
	if err != nil {
		return nil, fmt.Errorf("failed to get inactive students: %v", err)
	}

	var responses []models.StudentResponse
	for _, student := range students {
		studentWithRelations, err := s.studentRepo.GetStudentWithRelations(student.ID)
		if err != nil {
			continue
		}
		responses = append(responses, *s.toStudentResponse(studentWithRelations))
	}

	return responses, nil
}

func (s *studentService) ChangeStudentStatus(studentID uint, status int) error {
	// Check if student exists
	_, err := s.studentRepo.GetByID(studentID)
	if err != nil {
		return fmt.Errorf("student not found")
	}

	if err := s.studentRepo.UpdateStudentStatus(studentID, status); err != nil {
		return fmt.Errorf("failed to update student status: %v", err)
	}

	return nil
}

func (s *studentService) GetActiveStudents() ([]models.StudentResponse, error) {
	students, err := s.studentRepo.GetActiveStudents()
	if err != nil {
		return nil, fmt.Errorf("failed to get active students: %v", err)
	}

	var responses []models.StudentResponse
	for _, student := range students {
		studentWithRelations, err := s.studentRepo.GetStudentWithRelations(student.ID)
		if err != nil {
			continue
		}
		responses = append(responses, *s.toStudentResponse(studentWithRelations))
	}

	return responses, nil
}

func (s *studentService) GetInactiveStudents() ([]models.StudentResponse, error) {
	students, err := s.studentRepo.GetInactiveStudents()
	if err != nil {
		return nil, fmt.Errorf("failed to get inactive students: %v", err)
	}

	var responses []models.StudentResponse
	for _, student := range students {
		studentWithRelations, err := s.studentRepo.GetStudentWithRelations(student.ID)
		if err != nil {
			continue
		}
		responses = append(responses, *s.toStudentResponse(studentWithRelations))
	}

	return responses, nil
}

func (s *studentService) SearchStudents(searchTerm string, limit int, businessID ...uint) ([]models.StudentResponse, error) {
	students, err := s.studentRepo.SearchStudents(searchTerm, limit, businessID...)
	if err != nil {
		return nil, fmt.Errorf("failed to search students: %v", err)
	}

	var responses []models.StudentResponse
	for _, student := range students {
		studentWithRelations, err := s.studentRepo.GetStudentWithRelations(student.ID)
		if err != nil {
			continue
		}
		responses = append(responses, *s.toStudentResponse(studentWithRelations))
	}

	return responses, nil
}

func (s *studentService) SearchStudentsByBusiness(businessID uint, searchTerm string, limit int) ([]models.StudentResponse, error) {
	return s.SearchStudents(searchTerm, limit, businessID)
}

func (s *studentService) GetStudentStats(businessID ...uint) (map[string]interface{}, error) {
	return s.studentRepo.GetStudentStats(businessID...)
}

func (s *studentService) GetGuardianStats(businessID ...uint) (map[string]interface{}, error) {
	return s.studentRepo.GetGuardianStats(businessID...)
}

func (s *studentService) BulkUpdateStudentStatus(studentIDs []uint, status int) error {
	if len(studentIDs) == 0 {
		return fmt.Errorf("no student IDs provided")
	}

	if err := s.studentRepo.BulkUpdateStatus(studentIDs, status); err != nil {
		return fmt.Errorf("failed to bulk update student status: %v", err)
	}

	return nil
}

func (s *studentService) ValidateCreateStudentRequest(req models.CreateStudentRequest) error {
	if strings.TrimSpace(req.Name) == "" {
		return fmt.Errorf("name is required")
	}

	if req.UserID == 0 {
		return fmt.Errorf("user ID is required")
	}

	if req.BusinessID == 0 {
		return fmt.Errorf("business ID is required")
	}

	// Validate guardian email if provided
	if req.GuardianEmail != "" {
		// Basic email validation
		if !strings.Contains(req.GuardianEmail, "@") {
			return fmt.Errorf("invalid guardian email format")
		}
	}

	return nil
}

func (s *studentService) ValidateUpdateStudentRequest(req models.UpdateStudentRequest) error {
	if req.Name != "" && strings.TrimSpace(req.Name) == "" {
		return fmt.Errorf("name cannot be empty")
	}

	if req.Status != nil && (*req.Status < 0 || *req.Status > 1) {
		return fmt.Errorf("invalid status value")
	}

	// Validate guardian email if provided
	if req.GuardianEmail != "" && !strings.Contains(req.GuardianEmail, "@") {
		return fmt.Errorf("invalid guardian email format")
	}

	return nil
}

// Helper methods
func (s *studentService) toStudentResponse(student *models.Student) *models.StudentResponse {
	response := &models.StudentResponse{
		ID:             student.ID,
		Name:           student.Name,
		UserID:         student.UserID,
		BusinessID:     student.BusinessID,
		GuardianName:   student.GuardianName,
		GuardianNumber: student.GuardianNumber,
		GuardianEmail:  student.GuardianEmail,
		Information:    student.Information,
		Status:         student.Status,
		CreatedOn:      student.CreatedOn,
		UpdatedOn:      student.UpdatedOn,
	}

	// Add user details if loaded
	if student.User.ID != 0 {
		response.User = &models.UserResponse{
			ID:        student.User.ID,
			Name:      student.User.Name,
			Email:     student.User.Email,
			Role:      student.User.Role,
			Status:    student.User.Status,
			CreatedOn: student.User.CreatedOn,
		}
	}

	// Add business details if loaded
	if student.Business.ID != 0 {
		response.Business = &models.BusinessResponse{
			ID:        student.Business.ID,
			Name:      student.Business.Name,
			Slug:      student.Business.Slug,
			UserID:    student.Business.UserID,
			OwnerName: student.Business.OwnerName,
			Email:     student.Business.Email,
			Phone:     student.Business.Phone,
			Location:  student.Business.Location,
			Status:    student.Business.Status,
			CreatedOn: student.Business.CreatedOn,
		}
	}

	return response
}

File: frontend/src/services/studentService.ts
import { AxiosInstance } from 'axios';
import {
  Student,
  CreateStudentRequest,
  UpdateStudentRequest,
  StudentFilters,
  StudentStats,
  GuardianStats,
  StudentSearchResult,
  BulkUpdateStudentStatusRequest,
  StudentsResponse,
} from '@/types/student';

export class StudentService {
  private api: AxiosInstance;

  constructor(api: AxiosInstance) {
    this.api = api;
  }

  // Basic CRUD operations
  async getStudents(filters: StudentFilters = {}): Promise<StudentsResponse> {
    const params = new URLSearchParams();
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        params.append(key, value.toString());
      }
    });

    const response = await this.api.get(`/students?${params}`);
    return response.data;
  }

  async getStudent(id: number): Promise<Student> {
    const response = await this.api.get(`/students/${id}`);
    return response.data.data;
  }

  async createStudent(studentData: CreateStudentRequest): Promise<Student> {
    const response = await this.api.post('/students', studentData);
    return response.data.data;
  }

  async updateStudent(id: number, studentData: Partial<UpdateStudentRequest>): Promise<Student> {
    const response = await this.api.put(`/students/${id}`, studentData);
    return response.data.data;
  }

  async deleteStudent(id: number): Promise<void> {
    await this.api.delete(`/students/${id}`);
  }

  // Profile management (for student users)
  async getMyStudentProfile(): Promise<Student> {
    const response = await this.api.get('/my-student-profile');
    return response.data.data;
  }

  async updateMyStudentProfile(profileData: Partial<UpdateStudentRequest>): Promise<Student> {
    const response = await this.api.put('/my-student-profile', profileData);
    return response.data.data;
  }

  // Business-specific operations
  async getStudentsByBusiness(businessId: number, filters: StudentFilters = {}): Promise<StudentsResponse> {
    const params = new URLSearchParams();
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        params.append(key, value.toString());
      }
    });

    const response = await this.api.get(`/businesses/${businessId}/students?${params}`);
    return response.data;
  }

  async getActiveStudentsByBusiness(businessId: number): Promise<Student[]> {
    const response = await this.api.get(`/businesses/${businessId}/students/active`);
    return response.data.data;
  }

  async getInactiveStudentsByBusiness(businessId: number): Promise<Student[]> {
    const response = await this.api.get(`/businesses/${businessId}/students/inactive`);
    return response.data.data;
  }

  // Status operations
  async changeStudentStatus(id: number, status: number): Promise<void> {
    await this.api.patch(`/students/${id}/status`, { status });
  }

  async getActiveStudents(): Promise<Student[]> {
    const response = await this.api.get('/students/active');
    return response.data.data;
  }

  async getInactiveStudents(): Promise<Student[]> {
    const response = await this.api.get('/students/inactive');
    return response.data.data;
  }

  // Search functionality
  async searchStudents(query: string, limit: number = 10, businessId?: number): Promise<StudentSearchResult> {
    const params = new URLSearchParams();
    params.append('q', query);
    params.append('limit', limit.toString());
    if (businessId) {
      params.append('business_id', businessId.toString());
    }

    const response = await this.api.get(`/students/search?${params}`);
    return response.data.data;
  }

  // Statistics
  async getStudentStats(businessId?: number): Promise<StudentStats> {
    const params = new URLSearchParams();
    if (businessId) {
      params.append('business_id', businessId.toString());
    }

    const response = await this.api.get(`/students/stats?${params}`);
    return response.data.data;
  }

  async getGuardianStats(businessId?: number): Promise<GuardianStats> {
    const params = new URLSearchParams();
    if (businessId) {
      params.append('business_id', businessId.toString());
    }

    const response = await this.api.get(`/students/stats/guardians?${params}`);
    return response.data.data;
  }

  // Bulk operations
  async bulkUpdateStudentStatus(studentIds: number[], status: number): Promise<void> {
    await this.api.post('/students/bulk/status', {
      student_ids: studentIds,
      status,
    });
  }

  // Guardian-specific methods
  async getStudentsByGuardianEmail(email: string, businessId?: number): Promise<Student[]> {
    const filters: StudentFilters = {
      guardian_email: email,
      limit: 100,
    };

    if (businessId) {
      filters.business_id = businessId;
    }

    const response = await this.getStudents(filters);
    return response.data.students;
  }

  async getStudentsByGuardianName(name: string, businessId?: number): Promise<Student[]> {
    const filters: StudentFilters = {
      guardian_name: name,
      limit: 100,
    };

    if (businessId) {
      filters.business_id = businessId;
    }

    const response = await this.getStudents(filters);
    return response.data.students;
  }

  // Information management methods
  async updateStudentInformation(id: number, information: Record<string, any>): Promise<Student> {
    return this.updateStudent(id, { information });
  }

  async getStudentInformation(id: number): Promise<Record<string, any>> {
    const student = await this.getStudent(id);
    return student.information || {};
  }

  async addStudentInformationField(id: number, key: string, value: any): Promise<Student> {
    const student = await this.getStudent(id);
    const updatedInfo = { ...student.information, [key]: value };
    return this.updateStudentInformation(id, updatedInfo);
  }

  async removeStudentInformationField(id: number, key: string): Promise<Student> {
    const student = await this.getStudent(id);
    const updatedInfo = { ...student.information };
    delete updatedInfo[key];
    return this.updateStudentInformation(id, updatedInfo);
  }

  // Export/Import methods
  async exportStudents(filters: StudentFilters = {}, format: 'csv' | 'excel' = 'csv'): Promise<Blob> {
    const params = new URLSearchParams();
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        params.append(key, value.toString());
      }
    });
    params.append('format', format);

    const response = await this.api.get(`/students/export?${params}`, {
      responseType: 'blob'
    });
    return response.data;
  }

  async importStudents(file: File, businessId: number): Promise<{ success: number; errors: string[] }> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('business_id', businessId.toString());

    const response = await this.api.post('/students/import', formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return response.data;
  }

  // Validation methods
  async validateStudentUserExists(userId: number, excludeStudentId?: number): Promise<{ available: boolean }> {
    const params = new URLSearchParams();
    params.append('user_id', userId.toString());
    if (excludeStudentId) {
      params.append('exclude_id', excludeStudentId.toString());
    }

    const response = await this.api.get(`/students/validate-user?${params}`);
    return response.data;
  }

  async validateGuardianEmail(email: string): Promise<{ valid: boolean }> {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return { valid: emailRegex.test(email) };
  }

  // Helper methods for filtering and sorting
  getAvailableSortFields(): Array<{ value: string; label: string }> {
    return [
      { value: 'name', label: 'Student Name' },
      { value: 'guardian_name', label: 'Guardian Name' },
      { value: 'guardian_email', label: 'Guardian Email' },
      { value: 'guardian_number', label: 'Guardian Number' },
      { value: 'status', label: 'Status' },
      { value: 'created_on', label: 'Created Date' },
      { value: 'updated_on', label: 'Updated Date' },
    ];
  }

    getStatusOptions(): Array<{ value: number; label: string; color: string }> {
    return [
      { value: 1, label: 'Active', color: 'green' },
      { value: 0, label: 'Inactive', color: 'red' },
    ];
  }

  // Utility formatting methods
  formatGuardianContact(student: Student): string {
    const parts = [];
    if (student.guardian_name) parts.push(student.guardian_name);
    if (student.guardian_email) parts.push(student.guardian_email);
    if (student.guardian_number) parts.push(student.guardian_number);
    return parts.join(' • ') || 'No guardian info';
  }

  getStudentFullName(student: Student): string {
    return student.name || 'Unknown Student';
  }

  getStudentStatusLabel(status: number): string {
    return status === 1 ? 'Active' : 'Inactive';
  }

  getStudentStatusColor(status: number): string {
    return status === 1 ? 'green' : 'red';
  }

  hasGuardianContact(student: Student): boolean {
    return !!(student.guardian_email || student.guardian_number);
  }

  hasCompleteGuardianInfo(student: Student): boolean {
    return !!(student.guardian_name && (student.guardian_email || student.guardian_number));
  }

  // Information field management helpers
  getInformationFieldValue(student: Student, fieldKey: string, defaultValue: any = null): any {
    return student.information?.[fieldKey] ?? defaultValue;
  }

  hasInformationField(student: Student, fieldKey: string): boolean {
    return student.information && fieldKey in student.information;
  }

  getInformationFieldKeys(student: Student): string[] {
    return Object.keys(student.information || {});
  }

  // Contact validation helpers
  isValidEmail(email: string): boolean {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
  }

  isValidPhoneNumber(phone: string): boolean {
    // Basic phone validation - adjust regex based on your requirements
    const phoneRegex = /^[\+]?[1-9][\d]{0,15}$/;
    return phoneRegex.test(phone.replace(/[\s\-\(\)]/g, ''));
  }

  formatPhoneNumber(phone: string): string {
    // Basic phone formatting - customize based on your locale
    const cleaned = phone.replace(/\D/g, '');
    if (cleaned.length === 10) {
      return `(${cleaned.slice(0, 3)}) ${cleaned.slice(3, 6)}-${cleaned.slice(6)}`;
    }
    return phone;
  }

  // Filter helpers
  getStudentsWithGuardianEmail(students: Student[]): Student[] {
    return students.filter(student => student.guardian_email && student.guardian_email.trim());
  }

  getStudentsWithGuardianPhone(students: Student[]): Student[] {
    return students.filter(student => student.guardian_number && student.guardian_number.trim());
  }

  getStudentsWithIncompleteInfo(students: Student[]): Student[] {
    return students.filter(student => !this.hasCompleteGuardianInfo(student));
  }

  // Grouping helpers
  groupStudentsByStatus(students: Student[]): { active: Student[]; inactive: Student[] } {
    return students.reduce(
      (acc, student) => {
        if (student.status === 1) {
          acc.active.push(student);
        } else {
          acc.inactive.push(student);
        }
        return acc;
      },
      { active: [], inactive: [] } as { active: Student[]; inactive: Student[] }
    );
  }

  groupStudentsByGuardian(students: Student[]): Record<string, Student[]> {
    return students.reduce((acc, student) => {
      const guardianKey = student.guardian_email || student.guardian_name || 'No Guardian';
      if (!acc[guardianKey]) {
        acc[guardianKey] = [];
      }
      acc[guardianKey].push(student);
      return acc;
    }, {} as Record<string, Student[]>);
  }

  // Date helpers
  getStudentAge(student: Student, birthdateField: string = 'birthdate'): number | null {
    const birthdate = this.getInformationFieldValue(student, birthdateField);
    if (!birthdate) return null;

    const birth = new Date(birthdate);
    const today = new Date();
    let age = today.getFullYear() - birth.getFullYear();
    const monthDiff = today.getMonth() - birth.getMonth();

    if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birth.getDate())) {
      age--;
    }

    return age;
  }

  formatStudentCreatedDate(student: Student): string {
    return new Date(student.created_on).toLocaleDateString();
  }

  getStudentsByDateRange(students: Student[], startDate: Date, endDate: Date): Student[] {
    return students.filter(student => {
      const createdDate = new Date(student.created_on);
      return createdDate >= startDate && createdDate <= endDate;
    });
  }
}